File: app/controllers/assessment/QuestionReviewController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.assessment;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.text.PathProperties;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;
import java.util.stream.Collectors;
import models.base.GeneratedIdentityModel;
import models.exam.Exam;
import models.questions.Question;
import models.sections.ExamSectionQuestion;
import models.user.User;
import play.mvc.Http;
import play.mvc.Result;
import security.Authenticated;
import system.interceptors.Anonymous;
import validation.core.Attrs;

public class QuestionReviewController extends BaseController {

    private static final Exam.State[] VALID_STATES = {
        Exam.State.REVIEW,
        Exam.State.REVIEW_STARTED,
        Exam.State.GRADED,
        Exam.State.GRADED_LOGGED,
        Exam.State.REJECTED,
    };

    private boolean canAssess(User user, Exam exam) {
        return (
            exam.getParent().getExamOwners().contains(user) ||
            exam
                .getExamInspections()
                .stream()
                .anyMatch(ei -> ei.getUser().equals(user))
        );
    }

    @Authenticated
    @Restrict({ @Group("TEACHER") })
    @Anonymous(filteredProperties = { "user", "creator", "modifier" })
    public Result getEssays(Long examId, Optional<List<Long>> ids, Http.Request request) {
        Exam exam = DB.find(Exam.class, examId);
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (exam == null || !exam.isInspectedOrCreatedOrOwnedBy(user)) {
            return badRequest();
        }
        List<Long> questionIds = ids.orElse(Collections.emptyList());
        // This is the ordering of essay questions in the current exam
        List<Question> questionSequence = exam
            .getExamSections()
            .stream()
            .sorted()
            .flatMap(es -> es.getSectionQuestions().stream().sorted())
            .map(ExamSectionQuestion::getQuestion)
            .filter(question -> question.getType() == Question.Type.EssayQuestion)
            .filter(question -> questionIds.isEmpty() || questionIds.contains(question.getId()))
            .toList();

        // Ordered map for questions, have to take into account that answer's question is no longer found
        Map<Question, List<ExamSectionQuestion>> questionMap = createMapping(questionSequence);

        // All the answers for questions in this exam
        List<ExamSectionQuestion> answers = exam
            .getChildren()
            .stream()
            .filter(e -> canAssess(user, e))
            .filter(e -> Arrays.asList(VALID_STATES).contains(e.getState()))
            .flatMap(e -> e.getExamSections().stream())
            .flatMap(es -> es.getSectionQuestions().stream())
            .filter(esq -> esq.getQuestion().getType() == Question.Type.EssayQuestion)
            .filter(esq -> questionIds.isEmpty() || questionIds.contains(esq.getQuestion().getParent().getId()))
            .toList();

        // Get evaluation criteria from parent exam section questions
        Map<Question, String> evaluationCriteriaMap = exam
            .getExamSections()
            .stream()
            .flatMap(es -> es.getSectionQuestions().stream())
            .filter(esq -> esq.getQuestion().getType() == Question.Type.EssayQuestion)
            .filter(esq -> esq.getEvaluationCriteria() != null && esq.getQuestion() != null)
            .collect(
                Collectors.toMap(
                    ExamSectionQuestion::getQuestion,
                    ExamSectionQuestion::getEvaluationCriteria,
                    (existing, replacement) -> existing
                )
            );

        // Group essay answers by question and throw them in the ordered map
        questionMap.putAll(answers.stream().collect(Collectors.groupingBy(esq -> esq.getQuestion().getParent())));

        // Questions without answers, add separately because they aren't found through student exams
        questionSequence
            .stream()
            .filter(q -> !questionMap.containsKey(q))
            .forEach(q -> questionMap.put(q, Collections.emptyList()));

        // Pack as DTOs and serialize to JSON
        List<String> results = questionMap
            .entrySet()
            .stream()
            .map(e -> {
                Question key = e.getKey();
                String evaluationCriteria = evaluationCriteriaMap.get(key);
                return new QuestionEntry(e.getKey(), e.getValue(), evaluationCriteria).toJson();
            })
            .toList();

        String json = String.format("[%s]", String.join(", ", results));
        return writeAnonymousResult(request, ok(json).as("application/json"), exam.isAnonymous());
    }

    private static Map<Question, List<ExamSectionQuestion>> createMapping(List<Question> questions) {
        Comparator<Question> comparator = (o1, o2) -> {
            List<Long> ids = questions.stream().map(GeneratedIdentityModel::getId).toList();
            if (!ids.contains(o1.getId())) {
                return 1;
            }
            if (!ids.contains(o2.getId())) {
                return -1;
            }
            return ids.indexOf(o1.getId()) - ids.indexOf(o2.getId());
        };

        // Ordered map of questions into answers
        return new TreeMap<>(comparator);
    }

    // DTO
    private static class QuestionEntry {

        private final String question;
        private final List<String> answers;
        private final String evaluationCriteria;

        QuestionEntry(Question question, List<ExamSectionQuestion> answers, String evaluationCriteria) {
            PathProperties pp = PathProperties.parse(
                "(*, essayAnswer(attachment(*), *), question(parent(question), attachment(*), *), " +
                    "examSection(name, exam(id, hash, creator(id, email, userIdentifier, firstName, lastName), " +
                    "state, examInspections(user(id)))))"
            );
            this.question = DB.json().toJson(question, PathProperties.parse("(attachment(*), *)"));
            this.answers = answers
                .stream()
                .map(a -> DB.json().toJson(a, pp))
                .toList();
            this.evaluationCriteria = DB.json().toJson(evaluationCriteria);
        }

        private String toJson() {
            return String.format(
                "{\"question\": %s, \"answers\": %s, \"evaluationCriteria\": %s}",
                question,
                answers,
                evaluationCriteria
            );
        }
    }
}

File: app/controllers/assessment/QuestionReviewController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.assessment

import controllers.base.AnonymousHandler
import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.exam.Exam
import models.questions.Question
import models.sections.ExamSectionQuestion
import models.user.{Role, User}
import org.apache.pekko.stream.Materializer
import play.api.libs.json.{JsValue, Json, Writes}
import play.api.mvc.*
import security.scala.Auth.{AuthenticatedAction, authorized}
import security.scala.{Auth, AuthExecutionContext}
import system.interceptors.scala.AnonymousJsonFilter

import javax.inject.Inject
import scala.collection.immutable.TreeMap
import scala.concurrent.ExecutionContext
import scala.jdk.CollectionConverters.*

class QuestionReviewController @Inject() (
    val controllerComponents: ControllerComponents,
    val authenticated: AuthenticatedAction,
    val anonymous: AnonymousJsonFilter,
    implicit val ec: AuthExecutionContext,
    implicit val mat: Materializer
) extends BaseController
    with JavaApiHelper
    with AnonymousHandler
    with DbApiHelper:

  override protected def executionContext: ExecutionContext = ec
  override protected def materializer: Materializer         = mat

  case class QuestionEntry(question: JsValue, answers: Seq[JsValue], evaluationCriteria: JsValue)
  private object QuestionEntry:
    implicit val writes: Writes[QuestionEntry] = Json.writes[QuestionEntry]
    def apply(question: Question, answers: Seq[ExamSectionQuestion], evaluationCriteria: String): QuestionEntry =
      val answerPathProps = PathProperties.parse(
        """(*,
          |essayAnswer(attachment(*), *),
          |question(
          |  parent(question),
          |  attachment(*),
          |  *
          |),
          |examSection(name,
          |  exam(id, hash,
          |    creator(id, email, userIdentifier, firstName, lastName),
          |    state,
          |    examInspections(user(id))
          |  )
          |)
          |)""".stripMargin
      )
      val questionPathProps = PathProperties.parse("(attachment(*), *)")
      QuestionEntry(
        question = Json.parse(DB.json().toJson(question, questionPathProps)),
        answers = answers.map(a => Json.parse(DB.json().toJson(a, answerPathProps))),
        evaluationCriteria = Json.parse(DB.json().toJson(evaluationCriteria))
      )

  private val VALID_STATES =
    Seq(Exam.State.REVIEW, Exam.State.REVIEW_STARTED, Exam.State.GRADED, Exam.State.GRADED_LOGGED, Exam.State.REJECTED)

  def listEssays(examId: Long, ids: Option[List[Long]]): Action[AnyContent] = authenticated
    .andThen(authorized(Seq(Role.Name.TEACHER)))
    .andThen(anonymous(Set("user", "creator", "modifier"))) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      Option(DB.find(classOf[Exam], examId)) match
        case Some(exam) if exam.isInspectedOrCreatedOrOwnedBy(user) =>
          val questionIds = ids.getOrElse(List.empty)
          // This is the ordering of essay questions in the current exam
          val questionSequence = exam.getExamSections.asScala.toList.sorted
            .flatMap(_.getSectionQuestions.asScala.toList.sorted)
            .map(_.getQuestion)
            .filter(_.getType == Question.Type.EssayQuestion)
            .filter(q => questionIds.isEmpty || questionIds.contains(q.getId))

          val answers = exam.getChildren.asScala.toList
            .filter(canAssess(user, _))
            .filter(e => VALID_STATES.contains(e.getState))
            .flatMap(_.getExamSections.asScala)
            .flatMap(_.getSectionQuestions.asScala)
            .filter(_.getQuestion.getType == Question.Type.EssayQuestion)
            .filter(esq => questionIds.isEmpty || questionIds.contains(esq.getQuestion.getParent.getId))

          // Get evaluation criteria from parent exam section questions
          val evaluationCriteriaMap = exam.getExamSections.asScala.toList
            .flatMap(_.getSectionQuestions.asScala)
            .filter(esq =>
              esq.getQuestion.getType == Question.Type.EssayQuestion &&
                Option(esq.getEvaluationCriteria).isDefined &&
                Option(esq.getQuestion).isDefined
            )
            .distinctBy(_.getQuestion) // Keep first occurrence (like Java's merge function)
            .map(esq => esq.getQuestion -> esq.getEvaluationCriteria)
            .toMap

          val results = createMapping(answers, questionSequence)
            .map((question, answers) => QuestionEntry(question, answers, evaluationCriteriaMap.getOrElse(question, "")))
            .toSeq

          writeAnonymousResult(request, Ok(Json.toJson(results)), exam.isAnonymous)

        case _ => BadRequest
    }

  private def canAssess(user: User, exam: Exam) =
    exam.getParent.getExamOwners.contains(user) ||
      exam.getExamInspections.asScala.exists(_.getUser.equals(user))

  private def createMapping(
      answers: Seq[ExamSectionQuestion],
      questions: Seq[Question]
  ): Map[Question, List[ExamSectionQuestion]] =

    // Create ordering based on original question list
    val idToIndex = questions.map(_.getId).zipWithIndex.toMap
    implicit val questionOrdering: Ordering[Question] = Ordering.by { q =>
      idToIndex.getOrElse(q.getId, Int.MaxValue)
    }

    // Group essay answers by question parent
    val answersGrouped = answers
      .groupBy(_.getQuestion.getParent)
      .view
      .mapValues(_.toList)
      .toMap

    // Add questions without answers
    val questionsWithoutAnswers = questions
      .filterNot(answersGrouped.contains)
      .map(q => q -> List.empty[ExamSectionQuestion])
      .toMap

    // Combine and convert to a sorted map
    TreeMap.from(answersGrouped ++ questionsWithoutAnswers)
